pandas/_libs/lib.pyx

```diff
@@ -2876,24 +2875,54 @@ def map_infer_mask(
     mask : ndarray
         uint8 dtype ndarray indicating values not to apply `f` to.
     convert : bool, default True
-        Whether to call `maybe_convert_objects` on the resulting ndarray
+        Whether to call `maybe_convert_objects` on the resulting ndarray.
     na_value : Any, optional
         The result value to use for masked values. By default, the
-        input value is used
+        input value is used.
     dtype : numpy.dtype
         The numpy dtype to use for the result ndarray.
 
     Returns
     -------
     np.ndarray
     """
+    cdef Py_ssize_t n = len(arr)
+    result = np.empty(n, dtype=dtype)
+
+    _map_infer_mask(
+        result,
+        arr,
+        f,
+        mask,
+        convert,
+        na_value,
+        dtype,
+    )
+    if convert:
+        return maybe_convert_objects(result)
+    else:
+        return result
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _map_infer_mask(
```

Comment:
make this a cdef function?

Reply:
I don't believe so. When I try, I get:

```
Error compiling Cython file:
  ------------------------------------------------------------
  ...
      np.ndarray
      """
      cdef Py_ssize_t n = len(arr)
      result = np.empty(n, dtype=dtype)

      _map_infer_mask(
                    ^
  ------------------------------------------------------------

  /home/richard/dev/pandas/pandas/_libs/lib.pyx:2892:19: no suitable method found
```

I believe this is https://github.com/cython/cython/issues/2462

I'm not certain, but I believe it happens when calling a cdef function from a def function where the def function does _not_ use the fused types.